algorithms/src/snark/marlin/data_structures/proof.rs

```diff
@@ -46,108 +46,58 @@ pub struct Commitments<E: PairingEngine> {
-        Ok(())
-    }
-
-    fn uncompressed_size(&self) -> usize {
-        let mut size = 0;
-        size += self.witness_commitments.len() * CanonicalSerialize::uncompressed_size(&self.witness_commitments[0]);
-        size += CanonicalSerialize::uncompressed_size(&self.mask_poly);
-        size += CanonicalSerialize::uncompressed_size(&self.g_1);
-        size += CanonicalSerialize::uncompressed_size(&self.h_1);
-        size += CanonicalSerialize::uncompressed_size(&self.g_a);
-        size += CanonicalSerialize::uncompressed_size(&self.g_b);
-        size += CanonicalSerialize::uncompressed_size(&self.g_c);
-        size += CanonicalSerialize::uncompressed_size(&self.h_2);
-        size
-    }
-
-    fn deserialize<R: snarkvm_utilities::Read>(
+    fn deserialize_with_mode<R: snarkvm_utilities::Read>(
         batch_size: usize,
-        reader: &mut R,
+        mut reader: R,
+        compress: Compress,
+        validate: Validate,
     ) -> Result<Self, snarkvm_utilities::SerializationError> {
-        {
-            let mut witness_commitments = Vec::with_capacity(batch_size);
-            for _ in 0..batch_size {
-                witness_commitments.push(CanonicalDeserialize::deserialize(reader)?);
-            }
-            Ok(Commitments {
-                witness_commitments,
-                mask_poly: CanonicalDeserialize::deserialize(reader)?,
-                g_1: CanonicalDeserialize::deserialize(reader)?,
-                h_1: CanonicalDeserialize::deserialize(reader)?,
-                g_a: CanonicalDeserialize::deserialize(reader)?,
-                g_b: CanonicalDeserialize::deserialize(reader)?,
-                g_c: CanonicalDeserialize::deserialize(reader)?,
-                h_2: CanonicalDeserialize::deserialize(reader)?,
-            })
-        }
-    }
-
-    fn deserialize_uncompressed<R: snarkvm_utilities::Read>(
-        batch_size: usize,
-        reader: &mut R,
-    ) -> Result<Self, snarkvm_utilities::SerializationError> {
-        {
-            let mut witness_commitments = Vec::with_capacity(batch_size);
-            for _ in 0..batch_size {
-                witness_commitments.push(CanonicalDeserialize::deserialize_uncompressed(reader)?);
-            }
-            Ok(Commitments {
-                witness_commitments,
-                mask_poly: CanonicalDeserialize::deserialize_uncompressed(reader)?,
-                g_1: CanonicalDeserialize::deserialize_uncompressed(reader)?,
-                h_1: CanonicalDeserialize::deserialize_uncompressed(reader)?,
-                g_a: CanonicalDeserialize::deserialize_uncompressed(reader)?,
-                g_b: CanonicalDeserialize::deserialize_uncompressed(reader)?,
-                g_c: CanonicalDeserialize::deserialize_uncompressed(reader)?,
-                h_2: CanonicalDeserialize::deserialize_uncompressed(reader)?,
-            })
+        let mut witness_commitments = Vec::with_capacity(batch_size);
```

Comment:
```suggestion
        let mut witness_commitments = Vec::new();
```
this is not new to this PR, but we need to avoid `Vec::with_capacity`, otherwise we risk an overallocation panic in snarkOS; an alternative is to keep this call, but to add a capacity sanity check before it (to ensure that it's not too large)